Userland/Utilities/userdel.cpp

```diff
@@ -92,34 +114,54 @@ int main(int argc, char** argv)
     }
     endpwent();
 
-    if (fclose(temp_file)) {
+    setspent();
+    for (auto* spwd = getspent(); spwd; spwd = getspent()) {
+        if (strcmp(spwd->sp_namp, username)) {
+            if (putspent(spwd, temp_shadow_file) != 0) {
+                perror("failed to put an entry in the temporary shadow file");
+                return 1;
+            }
+        }
+    }
+    endspent();
+
+    if (fclose(temp_passwd_file)) {
         perror("fclose");
-        if (!rc)
-            rc = 1;
+        return 1;
+    }
+
+    if (fclose(temp_shadow_file)) {
+        perror("fclose");
+        return 1;
     }
 
-    if (rc == 0 && !user_exists) {
+    if (!user_exists) {
         warnln("specified user doesn't exist");
-        rc = 6;
+        return 1;
+    }
+
+    if (chmod(temp_passwd, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+        perror("chmod");
+        return 1;
     }
 
-    if (rc == 0 && chmod(temp_filename, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
+    if (chmod(temp_shadow, S_IRUSR | S_IWUSR | S_IRGRP | S_IROTH)) {
         perror("chmod");
-        rc = 1;
+        return 1;
     }
 
-    if (rc == 0 && rename(temp_filename, "/etc/passwd") < 0) {
+    if (rename(temp_passwd, "/etc/passwd") < 0) {
         perror("failed to rename the temporary passwd file");
-        rc = 1;
+        return 1;
     }
 
-    if (rc) {
-        if (unlink(temp_filename) < 0) {
-            perror("unlink");
-        }
-        return rc;
+    if (rename(temp_shadow, "/etc/shadow") < 0) {
+        perror("failed to rename the temporary shadow file");
+        return 1;
     }
 
+    unlink_temp_files_guard.disarm();
```

Comment:
```suggestion
    unlink_temp_files();
    unlink_temp_files_guard.disarm();
```

Reply:
I saw this in the original suggestion, however, the temp files are removed when they are renamed in the lines above so calling unlink_temp_files() here throws an unlink error.

<img width="598" alt="Screen Shot 2021-05-30 at 11 25 28 AM" src="https://user-images.githubusercontent.com/54456386/120110058-28b2c300-c13a-11eb-93f9-5d1259a85a39.png">

Reply:
oh, I missed that! thanks, all good then :)

Reply:
Thanks! :)